c10/cuda/CUDAStream.cpp

```diff
@@ -64,10 +64,13 @@ static LeakyStreamInternals default_streams[C10_COMPILE_TIME_MAX_GPUS];
 static std::once_flag device_flags[C10_COMPILE_TIME_MAX_GPUS];
 static std::atomic<uint32_t> low_priority_counters[C10_COMPILE_TIME_MAX_GPUS];
 static std::atomic<uint32_t> high_priority_counters[C10_COMPILE_TIME_MAX_GPUS];
+static std::atomic<uint32_t> external_counters[C10_COMPILE_TIME_MAX_GPUS];
 static std::array<LeakyStreamInternals, kStreamsPerPool>
     low_priority_streams[C10_COMPILE_TIME_MAX_GPUS];
 static std::array<LeakyStreamInternals, kStreamsPerPool>
     high_priority_streams[C10_COMPILE_TIME_MAX_GPUS];
+static std::array<LeakyStreamInternals, kStreamsPerPool>
```

Comment:
I am not fond of this, as in the current implementation it causes stream reusing when 32 streams are created on the same device without the user knowing, so the stream pools must be rewritten, or streams should be created and freed in an RAII way. However this should be all addressed together in a different PR.

Reply:
You should open a new issue with you nice wtf bug reproducer 😂

Reply:
It's pretty fun because this patch has the *exact* same limitations as the previous version of the patch LOL

Reply:
Is there a real bug here? 32 streams in pool limitation was chosen because hardware streams are limited in number, so even though, technically, you can create unbounded number of user facing soft streams, they'd still be backed by the same hardware streams.

Reply:
I think there are a few catches here. Disclaimer is I haven't looked into the detail of PyTorch's stream pool, so I am just applying common sense here 😅 

>> 32 streams in pool limitation was chosen because hardware streams are limited in number, so even though, technically, you can create unbounded number of user facing soft streams, they'd still be backed by the same hardware streams.
>
>  we expect there to be fewer than 32 actual hardware streams. 
1. I think the number of hardware streams is an implementation detail in CUDA/HIP's runtime, and users should not need to know nor would they notice (unless concurrency is a concern). But the current implementation seems to expose it in an unexpected way -- @emcastillo can say more about it.
2. A CUDA/HIP user always expects that if `cudaStreamCreate()` succeeds then the returned stream is valid. On a visual profiler (`nvvp` or `nsys`) they will see work done on each designated stream (the stream number can go far beyond 32), which reveals no information on the actual number of hardware streams, as expected.
3. 32 is an outdated number. On recent hardware it is at least 128 (see the first row in [this table](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#features-and-technical-specifications__technical-specifications-per-compute-capability)). But, again, it's implementation detail (and hardware dependent). Is there a good argument to hard-code it?

I am simply curious about the design choice and not really pushing for any change (either in this PR or a separate one).

Reply:
> One hacky temporary fix would be to prohibit more external streams than the cache size so there's never premature eviction. But a more durable fix would probably be to modify the CUDAStream class or {super, sub}class it to permit external streams and avoid a "pool" or cache. @ngimel and I don't think we want to modify PyTorch's existing stream pools, which have a very different usage pattern, and were designed to avoid constant stream creation and destruction. In this case, however, we do want the external streams to have a temporary lifetime.

I totally agree with this point. I will try to subclass it again,
I remember trying to do it originally but found some issues that I can't remember (most likely due to my own ineptitude)

Reply:
> 1. I think the number of hardware streams is an implementation detail in CUDA/HIP's runtime, and users should not need to know nor would they notice (unless concurrency is a concern). But the current implementation seems to expose it in an unexpected way -- @emcastillo can say more about it.

It is an implementation detail and this does not expose it. If a user does use 32+ user-facing streams then they could observe the stream pool round-robining the pool streams, however. We were concerned that this might confuse some users but in practice it doesn't seem to be an issue.

> 2. A CUDA/HIP user always expects that if `cudaStreamCreate()` succeeds then the returned stream is valid. On a visual profiler (`nvvp` or `nsys`) they will see work done on each designated stream (the stream number can go far beyond 32), which reveals no information on the actual number of hardware streams, as expected.
> 3. 32 is an outdated number. On recent hardware it is at least 128 (see the first row in [this table](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#features-and-technical-specifications__technical-specifications-per-compute-capability)). But, again, it's implementation detail (and hardware dependent). Is there a good argument to hard-code it?

The returned stream will always be valid. It's also important to note that the pool supports accessing 32 user-facing hardware streams but libraries like cuBLAS and cuDNN independently manage their own streams. If there's a real program suffering because PyTorch only supports 32 user-facing streams hardware streams then we could consider increasing the pool size.

> I am simply curious about the design choice and not really pushing for any change (either in this PR or a separate one).

The value might be outdated and we could consider increasing it on newer hardware, but we haven't heard that it's an issue. We could also look at alternative designs where streams were created as usual but then returned to the pool when they were about to be destroyed for reuse later.